File: LeetCode/Easy/594-Longest_Harmonious_Subsequence/594-LongestHarmoniousSubsequence.ex
defmodule Solution do
  @spec find_lhs(nums :: [integer]) :: integer
  def find_lhs(nums) do
    freq = nums
      |> Enum.reduce(%{}, fn num, acc ->
      Map.update(acc, num, 1, fn count -> count + 1 end)
    end)

    ans = 0

    Enum.reduce(Map.to_list(freq), ans, fn {k, v}, current_ans ->
      if Map.has_key?(freq, k + 1) do
        max(current_ans, v + Map.fetch!(freq, k + 1))
      else
        current_ans
      end
    end)
  end
end

File: LeetCode/Easy/594-Longest_Harmonious_Subsequence/594-LongestHarmoniousSubsequence.cpp
class Solution {
public:
    int findLHS(vector<int>& nums) {
        std::unordered_map<int, int> freq;
        int ans = 0;

        for (int num : nums) {
            freq[num]++;
        }

        for (const auto& [k, v] : freq) {
            if (freq.count(k+1))
                ans = max(ans, v + freq[k + 1]);
        }

        return ans;
    }
};